FlowCrypt/Functionality/Services/EKMService.swift

```diff
@@ -0,0 +1,188 @@
+//
+//  EKMService.swift
+//  FlowCrypt
+//
+//  Created by Ioan Moldovan on 4/28/22
+//  Copyright © 2017-present FlowCrypt a. s. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+protocol EKMServiceType {
+    func refreshKeysFromEKMIfNeeded(in viewController: UIViewController)
+}
+
+enum RefreshKeyError: Error {
+    case cancelPassPhrase
+}
+
+final class EKMService: EKMServiceType {
+
+    let appContext: AppContextWithUser
+    var askedUserPassPhrase = false
+    let keyMethods: KeyMethodsType
+
+    init(appContext: AppContextWithUser) {
+        self.appContext = appContext
+        self.keyMethods = KeyMethods()
+    }
+
+    func refreshKeysFromEKMIfNeeded(in viewController: UIViewController) {
+        Task {
+            do {
+                self.askedUserPassPhrase = false
+                let configuration = try await appContext.clientConfigurationService.configuration
+                guard configuration.checkUsesEKM() == .usesEKM else {
+                    return
+                }
+                let emailKeyManagerApi = EmailKeyManagerApi(clientConfiguration: configuration)
+                let idToken = try await IdTokenUtils.getIdToken(userEmail: appContext.user.email)
+                let keyDetails = try await emailKeyManagerApi.getPrivateKeys(idToken: idToken)
+                let localKeys = try appContext.encryptedStorage.getKeypairs(by: appContext.user.email)
+
+                var passPhrase = try fetchCurrentPassphrase()
+                let keysToUpdate = try findKeysToUpdate(from: keyDetails, localKeys: localKeys)
+
+                if keysToUpdate.isEmpty {
+                    return
+                }
+
+                for keyDetail in keysToUpdate {
+                    passPhrase = try await saveKeyToLocal(
+                        in: viewController,
+                        context: appContext,
+                        keyDetail: keyDetail,
+                        passPhrase: passPhrase
+                    )
```

Comment:
You no longer need to do this `passphrase = ` reassignment in each loop. Instead, ask for the pass phrase above the loop if it's missing. That way the logic is more obvious without inspecting the methods.

```
guard passPhrase else {
  passPhrase = try await requestPassPhraseWithModal(in: viewController)
} 
```

Then you don't need `askedUserPassPhrase` either. Less state is generally better, and it's best if code reads like prose from top to bottom, as events are happening. It makes for more readable code.